### PR TITLE
TCP docs, with lots of diagrams

### PR DESCRIPTION
**Pull request by rengolin**
While trying to simplify the TCP implementation to get the UDP one simpler, I  learned a lot about the TCPImpl and the reasons why it cannot (easily) be simplified beyond what it is.

These are my notes and diagrams on how the TCP implementation works and the RPC and Node connections interact with it.

I'm not sure how to share the rendered documents before merge, but I hosted locally and the diagrams render nicely, as I expected.

### CODE REVIEW

**Comment by jumaffre** [comment]
> I'm not sure how to share the rendered documents before merge, but I hosted locally and the diagrams render nicely, as I expected.

You can copy the rendered diagram and paste it in a comment on this PR ([example](https://github.com/microsoft/CCF/pull/3468)).

**Comment by rengolin** [comment]
State machine diagram:
image

**Comment by rengolin** [comment]
Server diagram:
image

**Comment by rengolin** [comment]
Client diagram:
image

**Review by achamayou — APPROVED** [review]

**Comment by achamayou** [comment]
@rengolin thank you for taking the time to document all of this so nicely, this is fantastic.

File: doc/overview/glossary.rst [hunk 1]
```diff
@@ -59,9 +62,15 @@ Glossary
   Ring Buffer
     The ring buffer is a data structure that allows communication between the (unprotected) host and the enclave. Data that is written to one side can be read on the other. Only specific types of messages are supported to make sure each package that goes across is read by the right process in the right way.
 
+  RPC
```
Comment (jumaffre) [inline doc/overview/glossary.rst:65]: Should `REST` be listed in the glossary too?

Comment (rengolin) [reply doc/overview/glossary.rst:65]: can do

**Pull request merged**